src/OrleansTestingHost/AppDomainSiloHost.cs

```diff
@@ -87,4 +101,87 @@ public void Shutdown()
             silo.Shutdown();
         }
     }
+
+    /// <summary>
+    /// Test hook functions for white box testing.
+    /// NOTE: this class has to and will be removed entirely. This requires the tests that currently rely on it, to assert using different mechanisms, such as with grains.
+    /// </summary>
+    internal class AppDomainTestHooks : MarshalByRefObject
+    {
+        private readonly Silo silo;
+
+        public AppDomainTestHooks(Silo silo)
+        {
+            this.silo = silo;
+        }
+
+        internal IBootstrapProvider GetBootstrapProvider(string name)
+        {
+            IBootstrapProvider provider = silo.BootstrapProviders.First(p => p.Name.Equals(name));
+            return CheckReturnBoundaryReference("bootstrap provider", provider);
+        }
+
+        /// <summary>Find the named storage provider loaded in this silo. </summary>
+        internal IStorageProvider GetStorageProvider(string name) => CheckReturnBoundaryReference("storage provider", (IStorageProvider)silo.StorageProviderManager.GetProvider(name));
+
+        private static T CheckReturnBoundaryReference<T>(string what, T obj) where T : class
+        {
+            if (obj == null) return null;
+            if (obj is MarshalByRefObject || obj is ISerializable)
+            {
+                // Referernce to the provider can safely be passed across app-domain boundary in unit test process
+                return obj;
+            }
+            throw new InvalidOperationException(
+                $"Cannot return reference to {what} {TypeUtils.GetFullName(obj.GetType())} if it is not MarshalByRefObject or Serializable");
+        }
+
+        public IDictionary<GrainId, IGrainInfo> GetDirectoryForTypeNamesContaining(string expr)
+        {
+            var x = new Dictionary<GrainId, IGrainInfo>();
+            foreach (var kvp in ((LocalGrainDirectory)silo.LocalGrainDirectory).DirectoryPartition.GetItems())
```

Comment:
In the original code before the move, it was using the private field that didn't need casting. Similarly these are test hooks that were just moved but not migrated yet, and should be migrated soon. Didn't make too much effort to make sure abstractions are kept in place, since they are going away.